pkg/zabbix_api_test.go

```diff
@@ -1 +1,232 @@
+		httpClient: NewTestClient(func(req *http.Request) *http.Response {
+			return &http.Response{
+				StatusCode: 500,
+				Body:       ioutil.NopCloser(bytes.NewBufferString(`{"result":"sampleResult"}`)),
+				Header:     make(http.Header),
+			}
+		}),
+		authToken: "sampleAuthToken",
+		logger:    hclog.Default(),
+	},
+}
+
+func TestZabbixAPIQuery(t *testing.T) {
+	resp, err := mockDataSource.ZabbixAPIQuery(context.Background(), mockDataSourceRequest(`{"target":{"method":"Method","params":{"param1" : "Param1"}}}`))
+
+	assert.Equal(t, "\"sampleResult\"", resp.GetResults()[0].GetMetaJson())
+	assert.Equal(t, "zabbixAPI", resp.GetResults()[0].GetRefId())
+	assert.Nil(t, err)
+}
+
+func TestZabbixAPIQueryEmptyQuery(t *testing.T) {
+	resp, err := mockDataSource.ZabbixAPIQuery(context.Background(), mockDataSourceRequest(``))
+
+	assert.Nil(t, resp)
+	assert.NotNil(t, err)
+}
+
+func TestZabbixAPIQueryNoQueries(t *testing.T) {
+	basicDatasourceRequest := &datasource.DatasourceRequest{
+		Datasource: &datasource.DatasourceInfo{
+			Id:   1,
+			Name: "TestDatasource",
+		},
+	}
+	resp, err := mockDataSource.ZabbixAPIQuery(context.Background(), basicDatasourceRequest)
+
+	assert.Nil(t, resp)
+	assert.Equal(t, "At least one query should be provided", err.Error())
+}
+
+func TestZabbixAPIQueryError(t *testing.T) {
+	resp, err := mockDataSourceError.ZabbixAPIQuery(context.Background(), mockDataSourceRequest(`{"target":{"method":"Method","params":{"param1" : "Param1"}}}`))
+
+	assert.Nil(t, resp)
+	assert.Equal(t, "ZabbixAPIQuery is not implemented yet", err.Error())
+}
+
+func TestLogin(t *testing.T) {
+	resp, err := mockDataSource.login(context.Background(), "apiURL", "username", "password")
+
+	assert.Equal(t, "sampleResult", resp)
+	assert.Nil(t, err)
+}
+
+func TestLoginError(t *testing.T) {
+	resp, err := mockDataSourceError.login(context.Background(), "apiURL", "username", "password")
+
+	assert.Equal(t, "", resp)
+	assert.NotNil(t, err)
+}
+
+func TestLoginWithDs(t *testing.T) {
+	resp, err := mockDataSource.loginWithDs(context.Background(), basicDatasourceInfo)
+
+	assert.Equal(t, "sampleResult", resp)
+	assert.Nil(t, err)
+}
+
+func TestLoginWithDsError(t *testing.T) {
+	resp, err := mockDataSourceError.loginWithDs(context.Background(), basicDatasourceInfo)
+
+	assert.Equal(t, "", resp)
+	assert.NotNil(t, err)
+}
+
+func TestZabbixRequest(t *testing.T) {
+	resp, err := mockDataSource.ZabbixRequest(context.Background(), basicDatasourceInfo, "method", make(map[string]interface{}))
+	assert.Equal(t, "sampleResult", resp.MustString())
+	assert.Nil(t, err)
+}
+
+func TestZabbixRequestWithNoAuthToken(t *testing.T) {
+	var mockDataSource = mockZabbixDatasource{
+		ZabbixDatasource{
+			queryCache: NewCache(10*time.Minute, 10*time.Minute),
+			httpClient: NewTestClient(func(req *http.Request) *http.Response {
+				return &http.Response{
+					StatusCode: 200,
+					Body:       ioutil.NopCloser(bytes.NewBufferString(`{"result":"auth"}`)),
```

Comment:
Is this supposed to test cases where auth is required? I would expect it to return a 400 error if so, and to somehow check that we make a call to get the auth token and try again.